File: android/lib/core/crypto/crypto_engine.dart
import 'dart:io';
import 'dart:typed_data';
import 'package:cryptography/cryptography.dart';

/// معادل CryptoEngine.cs
/// رمزنگاری/رمزگشایی chunk به chunk با XChaCha20-Poly1305 (ساختار IETF ترکیبی:
/// ciphertext + 16 بایت tag پشت سرهم، دقیقاً مثل خروجی Sodium.Core در نسخه‌ی ویندوز)
class CryptoEngine {
  static const int defaultChunkSize = 4 * 1024 * 1024; // 4MB
  static const int _macLength = 16;

  static final _aead = Xchacha20.poly1305Aead();

  // nonce هر chunk = baseNonce با 8 بایت آخر XOR شده با شماره chunk (little-endian)
  static Uint8List _deriveChunkNonce(Uint8List baseNonce, int chunkIndex) {
    final chunkNonce = Uint8List.fromList(baseNonce);
    final indexBytes = ByteData(8)..setInt64(0, chunkIndex, Endian.little);
    for (int i = 0; i < 8; i++) {
      chunkNonce[baseNonce.length - 8 + i] ^= indexBytes.getUint8(i);
    }
    return chunkNonce;
  }

  static Future<void> encryptStream(
    RandomAccessFile input,
    RandomAccessFile output,
    Uint8List key,
    Uint8List baseNonce, {
    int chunkSize = defaultChunkSize,
    void Function(int processedBytes)? onProgress,
  }) async {
    final secretKey = SecretKeyData(key);
    int chunkIndex = 0;
    int totalRead = 0;

    while (true) {
      final chunk = await input.read(chunkSize);
      if (chunk.isEmpty) break;

      final chunkNonce = _deriveChunkNonce(baseNonce, chunkIndex);
      final secretBox = await _aead.encrypt(
        chunk,
        secretKey: secretKey,
        nonce: chunkNonce,
      );

      // ciphertext + tag پشت سر هم، همون فرمتی که ویندوز می‌نویسه
      final encrypted = Uint8List(secretBox.cipherText.length + _macLength);
      encrypted.setRange(0, secretBox.cipherText.length, secretBox.cipherText);
      encrypted.setRange(
          secretBox.cipherText.length, encrypted.length, secretBox.mac.bytes);

      final lengthBytes = ByteData(4)
        ..setInt32(0, encrypted.length, Endian.little);
      await output.writeFrom(lengthBytes.buffer.asUint8List());
      await output.writeFrom(encrypted);

      // پیشرفت بر اساس حجم خام خونده‌شده از فایل ورودی (همون واحدی که
      // totalSize در EncryptionService بر مبنای اون حساب میشه)
      totalRead += chunk.length;
      chunkIndex++;
      onProgress?.call(totalRead);
    }
  }

  static Future<void> decryptStream(
    RandomAccessFile input,
    RandomAccessFile output,
    Uint8List key,
    Uint8List baseNonce, {
    void Function(int processedBytes)? onProgress,
  }) async {
    final secretKey = SecretKeyData(key);
    int chunkIndex = 0;
    int totalRead = 0;

    while (true) {
      final lengthBuffer = await input.read(4);

      // ۰ بایت خوندیم = پایان طبیعی فایل
      if (lengthBuffer.isEmpty) break;

      // کمتر از ۴ بایت = فایل ناقص/آسیب‌دیده (نصفه کپی شده و غیره)
      if (lengthBuffer.length < 4) {
        throw const FormatException(
            'فایل ناقص یا آسیب‌دیده است (هدر chunk ناقص).');
      }

      final chunkLength =
          ByteData.sublistView(lengthBuffer).getInt32(0, Endian.little);

      // 🆕 سقف منطقی: بزرگ‌ترین chunk مجاز در هر دو پلتفرم + تگ MAC (جلوگیری از DoS)
      const maxChunkLength = 64 * 1024 * 1024;
      if (chunkLength < _macLength || chunkLength > maxChunkLength) {
        throw const FormatException('فایل آسیب‌دیده است (طول chunk نامعتبر).');
      }

      final encryptedChunk = await input.read(chunkLength);
      if (encryptedChunk.length < chunkLength) {
        throw const FormatException('فایل ناقص یا آسیب‌دیده است (chunk ناقص).');
      }

      final chunkNonce = _deriveChunkNonce(baseNonce, chunkIndex);
      final cipherText = encryptedChunk.sublist(0, chunkLength - _macLength);
      final macBytes = encryptedChunk.sublist(chunkLength - _macLength);

      List<int> decrypted;
      try {
        final secretBox = SecretBox(
          cipherText,
          nonce: chunkNonce,
          mac: Mac(macBytes),
        );
        decrypted = await _aead.decrypt(secretBox, secretKey: secretKey);
      } on SecretBoxAuthenticationError {
        throw const FormatException('رمز عبور اشتباه است یا فایل آسیب دیده.');
      } catch (e) {
        // خطاهای دیگه (مثلاً کمبود حافظه حین decrypt) رو هم واضح گزارش بده
        throw FormatException('خطای رمزگشایی: $e');
      }

      await output.writeFrom(decrypted);

      // نکته‌ی مهم: بر اساس حجم decrypted شده جمع می‌زنیم (نه chunkLength
      // رمزشده‌ی روی دیسک که ۱۶ بایت tag رو هم داخلش داره) تا با totalSize
      // (که در DecryptionService بر مبنای حجم واقعی payload حساب میشه)
      // هماهنگ باشه و درصد پیشرفت دقیق‌تر بشه.
      totalRead += decrypted.length;
      chunkIndex++;
      onProgress?.call(totalRead);
    }
  }
}

File: windows/SFE.Core/Crypto/CryptoEngine.cs
using Sodium;
using System.Buffers.Binary;
using System.Security.Cryptography;

namespace SFE.Core.Crypto;

public static class CryptoEngine
{
    private const int ChunkSize = 4 * 1024 * 1024; // 4MB

    // ── هر chunk یک nonce منحصربه‌فرد میگیره ──────────
    // nonce chunk = XOR(baseNonce, chunkCounter)
    // این روش استاندارد و امن هست
    private static byte[] DeriveChunkNonce(byte[] baseNonce, long chunkIndex)
    {
        var chunkNonce = (byte[])baseNonce.Clone();
        var indexBytes = new byte[8];
        BinaryPrimitives.WriteInt64LittleEndian(indexBytes, chunkIndex);
        for (int i = 0; i < 8; i++)
            chunkNonce[baseNonce.Length - 8 + i] ^= indexBytes[i];
        return chunkNonce;
    }

    public static void EncryptStream(
        Stream inputStream,
        Stream outputStream,
        byte[] key,
        byte[] baseNonce,
        int chunkSize = ChunkSize,
        IProgress<long>? progress = null,
        CancellationToken cancellationToken = default)
    {
        var buffer = new byte[chunkSize];
        long totalBytesRead = 0;
        long chunkIndex = 0;
        int bytesRead;
        while ((bytesRead = inputStream.Read(buffer, 0, buffer.Length)) > 0)
        {
            cancellationToken.ThrowIfCancellationRequested();
            var chunkNonce = DeriveChunkNonce(baseNonce, chunkIndex);
            var chunk = buffer[..bytesRead];
            var encrypted = SecretAeadXChaCha20Poly1305.Encrypt(chunk, chunkNonce, key);
            CryptographicOperations.ZeroMemory(chunkNonce);
            var lengthBytes = BitConverter.GetBytes(encrypted.Length);
            outputStream.Write(lengthBytes, 0, 4);
            outputStream.Write(encrypted, 0, encrypted.Length);
            totalBytesRead += bytesRead;
            chunkIndex++;
            progress?.Report(totalBytesRead);
        }
    }

    public static void DecryptStream(
        Stream inputStream,
        Stream outputStream,
        byte[] key,
        byte[] baseNonce,
        IProgress<long>? progress = null,
        CancellationToken cancellationToken = default)
    {
        long totalBytesRead = 0;
        long chunkIndex = 0;
        var lengthBuffer = new byte[4];
        while (inputStream.Read(lengthBuffer, 0, 4) == 4)
        {
            cancellationToken.ThrowIfCancellationRequested();
            var chunkLength = BitConverter.ToInt32(lengthBuffer, 0);

            // 🆕 اعتبارسنجی طول chunk (جلوگیری از تخصیص حافظه با فایل مخرب/خراب)
                        if (chunkLength <= 0 || chunkLength > 64 * 1024 * 1024)
                throw new InvalidDataException("ساختار فایل نامعتبر است.");

            var encryptedChunk = new byte[chunkLength];
            inputStream.ReadExactly(encryptedChunk, 0, chunkLength);

            var chunkNonce = DeriveChunkNonce(baseNonce, chunkIndex);
            byte[] decrypted;
            try
            {
                decrypted = SecretAeadXChaCha20Poly1305.Decrypt(
                    encryptedChunk, chunkNonce, key);
            }
            catch
            {
                CryptographicOperations.ZeroMemory(chunkNonce);
                throw new InvalidDataException("رمز عبور اشتباه است یا فایل آسیب دیده.");
            }
            CryptographicOperations.ZeroMemory(chunkNonce);
            outputStream.Write(decrypted, 0, decrypted.Length);
            totalBytesRead += chunkLength;
            chunkIndex++;
            progress?.Report(totalBytesRead);
        }
    }
}